cpp/src/column/column.cu

```diff
@@ -275,6 +278,123 @@ struct create_column_from_view_vector {
 
 };
 
+template <typename T>
+struct partition_map_fn {
+  size_type const* offsets_ptr;
+  size_type const* partition_ptr;
+  T const** data_ptr;
+
+  T __device__ operator()(size_type i) {
+    auto const partition = partition_ptr[i];
+    auto const offset = offsets_ptr[partition];
+    return data_ptr[partition][i - offset];
+  }
+};
+
+template <typename T>
+struct binary_search_fn {
+  size_type const* offsets_ptr;
+  size_t const input_size;
+  T const** data_ptr;
+
+  T __device__ operator()(size_type i) {
+    auto const offset_it = thrust::upper_bound(thrust::seq,
+      offsets_ptr, offsets_ptr + input_size, i) - 1;
+    auto const partition = offset_it - offsets_ptr;
+    auto const offset = *offset_it;
+    return data_ptr[partition][i - offset];
+  }
+};
+
+auto compute_partition_map(thrust::device_vector<size_type> const& d_offsets,
+    size_t output_size, cudaStream_t stream) {
+
+  thrust::device_vector<size_type> d_partition_map(output_size);
+
+  // Scatter 1s at the start of each partition, then scan to fill the map
+  auto const const_it = thrust::make_constant_iterator(size_type{1});
+  thrust::scatter(rmm::exec_policy(stream)->on(stream),
+    const_it, const_it + (d_offsets.size() - 2), // skip first and last offset
+    std::next(d_offsets.cbegin()), // skip first offset, leaving it set to 0
+    d_partition_map.begin());
+
+  thrust::inclusive_scan(rmm::exec_policy(stream)->on(stream),
+    d_partition_map.cbegin(), d_partition_map.cend(),
+    d_partition_map.begin());
+
+  return d_partition_map;
+}
+
+// Allow strategy switching at runtime for easier benchmarking
+// TODO remove when done
+static concatenate_mode current_mode = concatenate_mode::UNOPTIMIZED;
+void temp_set_concatenate_mode(concatenate_mode mode) {
+  current_mode = mode;
+}
+
+struct optimized_concatenate {
+  template <typename T,
+      std::enable_if_t<is_fixed_width<T>()>* = nullptr>
+  std::unique_ptr<column> operator()(
+      std::vector<column_view> const& views,
+      rmm::mr::device_memory_resource* mr,
+      cudaStream_t stream) {
+
+    // Compute the partition offsets
+    thrust::host_vector<size_type> offsets(views.size() + 1);
+    // NOTE tried to use thrust::transform_inclusive_scan here, but got
+    // error: __host__ called from __host__ __device__
+    thrust::transform(views.cbegin(), views.cend(), std::next(offsets.begin()),
+      [](column_view const& col) {
+        return col.size();
```

Comment:
This lambda was causing problems with `thrust::transform_inclusive_scan` and `thrust::make_transform_iterator`, but it works fine with `thrust::transform`

Reply:
Try just doing an `inclusive_scan` over a transform iterator? That's basically what `thrust::transform` does, so maybe it won't change anything.

Reply:
When I used `inclusive_scan` with a `transform_iterator`, I got the same error that I did with `transform_inclusive_scan`